modules/dnn/src/ocl4dnn/src/ocl4dnn_pool.cpp

```diff
@@ -0,0 +1,202 @@
+    }
+
+    kernel_h_ = kernel_shape_[0];
+    kernel_w_ = kernel_shape_[1];
+    stride_h_ = stride_[0];
+    stride_w_ = stride_[1];
+    pad_h_ = pad_[0];
+    pad_w_ = pad_[1];
+    height_ = im_in_shape_[0];
+    width_ = im_in_shape_[1];
+    pooled_height_ = im_out_shape_[0];
+    pooled_width_ = im_out_shape_[1];
+
+    count_ = 1;
+    for (int32_t i = 0; i < config.out_shape.size(); ++i)
+    {
+        count_ *= config.out_shape[i];
+    }
+}
+
+template<typename Dtype>
+OCL4DNNPool<Dtype>::~OCL4DNNPool()
+{
+    mask_idx_.release();
+}
+
+template<typename Dtype>
+bool OCL4DNNPool<Dtype>::Forward(const UMat& bottom,
+                                 UMat& top,
+                                 UMat& top_mask)
+{
+    bool ret = true;
+    ocl::Queue queue = ocl::Queue::getDefault();
+    size_t global[] = { 128 * 128 };
+    size_t local[] = { 128 };
+    cl_uint argIdx = 0;
+
+    // support 2D case
+    switch (pool_method_)
+    {
+    case LIBDNN_POOLING_METHOD_MAX:
+        {
+            if (top_mask.empty() && mask_idx_.empty())
+            {
+                mask_idx_.create(1, count_, CV_32FC1);
+            }
+            ocl::Kernel oclk_max_pool_forward(CL_KERNEL_SELECT("max_pool_forward"),
+                                              cv::ocl::dnn::ocl4dnn_pooling_oclsrc);
+
+            argIdx = 0;
+            oclk_max_pool_forward.set(argIdx++, count_);
+            oclk_max_pool_forward.set(argIdx++, (cl_mem) bottom.handle(ACCESS_READ));
+            oclk_max_pool_forward.set(argIdx++, batch_size_);
+            oclk_max_pool_forward.set(argIdx++, channels_);
+            oclk_max_pool_forward.set(argIdx++, height_);
+            oclk_max_pool_forward.set(argIdx++, width_);
+            oclk_max_pool_forward.set(argIdx++, pooled_height_);
+            oclk_max_pool_forward.set(argIdx++, pooled_width_);
+            oclk_max_pool_forward.set(argIdx++, kernel_h_);
+            oclk_max_pool_forward.set(argIdx++, kernel_w_);
+            oclk_max_pool_forward.set(argIdx++, stride_h_);
+            oclk_max_pool_forward.set(argIdx++, stride_w_);
+            oclk_max_pool_forward.set(argIdx++, pad_h_);
+            oclk_max_pool_forward.set(argIdx++, pad_w_);
+            oclk_max_pool_forward.set(argIdx++, (cl_mem) top.handle(ACCESS_WRITE));
+            oclk_max_pool_forward.set(argIdx++, mask_idx_.empty() ? 0 : 1);
+            oclk_max_pool_forward.set(argIdx++, mask_idx_.empty() ? NULL :
+                                      (cl_mem) mask_idx_.handle(ACCESS_WRITE));
+            oclk_max_pool_forward.set(argIdx++, (cl_mem) top_mask.handle(ACCESS_WRITE));
+
+            oclk_max_pool_forward.run(1, global, local, false);
+        }
+        break;
+    case LIBDNN_POOLING_METHOD_AVE:
+        {
+            ocl::Kernel oclk_ave_pool_forward(CL_KERNEL_SELECT("ave_pool_forward"),
+                                              cv::ocl::dnn::ocl4dnn_pooling_oclsrc);
+
```

Comment:
Error checks are mandatory (see MacOS builds):
```
if (kernel.empty())
    return false;
```

Reply:
Fixed in the update patch. Checks on ocl4dnn kernel create and run are added.